plugins/catalog/src/state/reducer.ts

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Entity } from '@backstage/catalog-model';
+import { collectTags, entityPredicate } from './filters';
+import { EntityFilter, EntityListState } from './types';
+
+export const entityListInitialState: EntityListState = {
```

Comment:
so i've been wondering what it would be like if the state would hold essentially

```ts
filters: {
  [filterId: string]: {
    predicate?: (entity: Entity) => boolean;
    serverFilters?: Record<string, string[]>;
  }
}
loading: boolean;
error: Error;
serverFilteredEntities: Entity[];
clientFilteredEntities: Entity[];
```

whenever anybody changes their server filters, the loading cycle starts for the server filters. Whenever the client filters change, the server filtered ones are additionally client filtered.

This would mean that a component would carry the responsibility for maintaining what it needs in terms of being able to supply those server/client filters.

Pseudocode:

```tsx
const StarredFilter = () => {
  const { starred } = useStarred();
  const [enabled, setEnabled] = useState<boolean>(false);
  const { loading, serverFilteredEntities, setFilters } = useFilterContext('starred');

  useEffect(() => {
    setFilters({
      predicate: e => starred.has(stringifyEntityRef(e))
    });
  }, [starred, enabled]);

  const matches = serverFilteredEntities.filter(e => starred.has(stringifyEntityRef(e))).length;
  return (
    <Checkbox disabled={loading} selected={enabled} onChange={() => setEnabled(e => !e)} /> {matches}
  );
};
```

that was totally just a shot in the dark but maybe it conveys the idea well enough :)

Reply:
Yep we could try to do some splitting perhaps, but I'm concerned that we'd need like two contexts where one of them depends on the other, or something like that. As long as it from the _caller_ (component) side is just `setFilter -> entities[]` effectively, I am not that concerned that the context holds both types of filtering together, since they depend strongly on each other.

Reply:
I was originally thinking very very roughly around something like this albeit I haven't thought through the code structure:

```ts
// additional filters are passed down through a top level prop and/or apiRef from the index page
const useEntityListState = (additionalFilters?: EntityFilter[]) => {
  // merges the additional filter types with the base set
}

// Components apply a filter action
applyFilter('build-in-type', [params])
applyFilter('custom-type', [custom-params])

// Reducer logic
// - get full list of applied filters
// - create 2 groups based on if the filter is a client filer or server filter
// - apply the filter groups in a chain of state change side effects which fetches server entities and then applies the client side ones to that result.
```

For this to work, I think the `applyFilter` action will need to be `thunk` probably to handle the side effects. I'm currently missing the mental gap to tie reducer / redux style side affects into the blend of hooks + useReducer here.